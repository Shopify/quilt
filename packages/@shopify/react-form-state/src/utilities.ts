import isEqual from 'fast-deep-equal';

export {isEqual};

export function mapObject<Input, Output>(
  input: Input,
  mapper: (value: any, key: keyof Input) => any,
) {
  return Object.entries(input).reduce((accumulator: any, [key, value]) => {
    accumulator[key] = mapper(value, key as keyof Input);
    return accumulator;
  }, {}) as Output;
}

export function push<T>(array: T[], ...values: T[]) {
  return array.concat(values);
}

export function remove<T>(array: T[], targetIndex: number) {
  return array.filter((_, index) => index !== targetIndex);
}

export function replace<T>(array: T[], targetIndex: number, newValue: T) {
  return array.map((value, index) => {
    if (index !== targetIndex) {
      return value;
    }

    return newValue;
  });
}

export function set<InputType extends object>(
  rootObject: InputType,
  path: string[],
  value: any,
) {
  if (path.length === 0) {
    return rootObject;
  } else if (path.length === 1) {
    return {
      ...(rootObject as any),
      [path[0]]: value,
    };
  } else {
    const [current, ...rest] = path;

    return {
      ...(rootObject as any),
      [current]: set(rootObject[current] || {}, rest, value),
    } as InputType;
  }
}

export function flatMap<T>(
  array: any[],
  mapper: (item: any, index?: number) => T | T[],
): T[] {
  return array.reduce(
    (acc, item, index) => acc.concat(mapper(item, index)),
    [],
  );
}
