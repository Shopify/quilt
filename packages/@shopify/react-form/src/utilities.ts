import {ChangeEvent} from 'react';

import get from 'get-value';

import {
  Validates,
  Validator,
  FieldOutput,
  FieldDictionary,
  Field,
  FormError,
} from './types';

export function isField<T extends object>(
  input: FieldOutput<T>,
): input is Field<T> {
  return (
    Object.prototype.hasOwnProperty.call(input, 'value') &&
    Object.prototype.hasOwnProperty.call(input, 'onChange') &&
    Object.prototype.hasOwnProperty.call(input, 'onBlur') &&
    Object.prototype.hasOwnProperty.call(input, 'defaultValue')
  );
}

export function mapObject<Output>(
  input: any,
  mapper: (value: any, key: any) => any,
) {
  return Object.keys(input).reduce((accumulator: any, key) => {
    const value = input[key];
    accumulator[key] = mapper(value, key);
    return accumulator;
  }, {}) as Output;
}

export function normalizeValidation<Value, Context extends object = {}>(
  input: Validates<Value, Context>,
): Validator<Value, Context>[] {
  return Array.isArray(input) ? input : [input];
}

export function isChangeEvent(
  value: any,
): value is ChangeEvent<HTMLInputElement> {
  return (
    typeof value === 'object' &&
    Reflect.has(value, 'target') &&
    Reflect.has(value.target, 'value')
  );
}

export function propagateErrors(
  fieldBag: {[key: string]: FieldOutput<any>},
  errors: FormError[],
) {
  errors.forEach(error => {
    if (error.field == null) {
      return;
    }

    const got = get(fieldBag, error.field);

    if (isField(got)) {
      if (got.error !== error.message) {
        got.setError(error.message);
      }
    }
  });
}

export function validateAll(fieldBag: {[key: string]: FieldOutput<any>}) {
  const fields = Object.values(fieldBag);
  const errors: FormError[] = [];

  function validate(field: Field<unknown>) {
    const message = field.runValidation();
    if (message) {
      errors.push({message});
    }
  }

  function validateDictionary(fields: FieldDictionary<any>) {
    Object.values(fields).forEach(validate);
  }

  for (const item of fields) {
    if (isField(item)) {
      validate(item);
    } else if (Array.isArray(item)) {
      item.map(validateDictionary);
    } else {
      validateDictionary(item);
    }
  }

  return errors;
}

export function noop() {}
