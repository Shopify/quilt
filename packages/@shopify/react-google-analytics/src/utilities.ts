// We want only the final segment of a domain, without any
// preceding subdomains. However, we do keep leading `.` when
// subdomains exist.
export function getRootDomain(domain: string) {
  const parts = domain.split('.');
  const baseDomainOnly = parts.slice(-2).join('.');

  return parts.length > 2 ? `.${baseDomainOnly}` : baseDomainOnly;
}

export function noop() {}
