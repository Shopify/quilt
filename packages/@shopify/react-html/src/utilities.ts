export const SERIALIZE_ATTRIBUTE = 'data-serialized-id';
export const MANAGED_ATTRIBUTE = 'data-react-html';

export function getSerializationsFromDocument() {
  const serializations = new Map<string, unknown>();

  if (typeof document === 'undefined') {
    return serializations;
  }

  for (const node of document.querySelectorAll(`[${SERIALIZE_ATTRIBUTE}]`)) {
    serializations.set(
      node.getAttribute(SERIALIZE_ATTRIBUTE)!,
      getSerializedFromNode(node),
    );
  }

  return serializations;
}

function getSerializedFromNode<Data>(node: Element): Data | undefined {
  const value = node.textContent;

  try {
    return value ? JSON.parse(value) : undefined;
  } catch {
    return undefined;
  }
}

export function getSerialized<Data>(id: string) {
  const node = document.querySelector(`[${SERIALIZE_ATTRIBUTE}="${id}"]`);

  if (node == null) {
    throw new Error(`No serializations found for id "${id}"`);
  }

  return getSerializedFromNode<Data>(node);
}

// We hide the document in development by default in order to prevent
// flashes of unstyled content. This function will show the document
// after the styles have loaded.
export function showPage(): Promise<void> {
  // eslint-disable-next-line no-process-env
  return process.env.NODE_ENV === 'development' &&
    typeof document !== 'undefined'
    ? new Promise(resolve => {
        setTimeout(() => {
          document.body.style.visibility = '';
          resolve();
        }, 0);
      })
    : Promise.resolve();
}
