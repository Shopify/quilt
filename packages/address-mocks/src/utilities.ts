import {fetch} from '@shopify/jest-dom-mocks';
import type {
  LoadCountriesResponse,
  LoadCountryResponse,
  ResponseError,
} from '@shopify/address-consts';
import {GRAPHQL_ENDPOINT} from '@shopify/address-consts';

import type {Fixtures, MockOptions} from './types';

const OPERATION_NAMES = ['countries', 'country'];

type FixtureLoader = () => Fixtures | Promise<Fixtures>;

export const defaultFixtureLoader: FixtureLoader = async () => {
  return (await import('./fixtures/fixtures-light')).default;
};

export function mockCountryRequestsLazy(
  loadFixtures: FixtureLoader = defaultFixtureLoader,
  options: MockOptions = {},
) {
  return fetch.post(requestMatcher, createMockHandler(loadFixtures), {
    overwriteRoutes: false,
    ...options,
  });
}

export function requestMatcher(url, options) {
  if (url !== GRAPHQL_ENDPOINT || typeof options.body !== 'string') {
    return false;
  }

  const body = JSON.parse(options.body);
  return OPERATION_NAMES.includes(body.operationName);
}

export function createMockHandler(loadFixtures: FixtureLoader) {
  let fixtures: Fixtures;
  return async function lazyRequestHandler(
    _,
    {body},
  ): Promise<LoadCountriesResponse | LoadCountryResponse | ResponseError> {
    if (!fixtures) {
      fixtures = await loadFixtures();
    }
    const locales = Object.keys(fixtures.countries);
    const {variables, operationName} = JSON.parse(body as string);
    const {countryCode, locale, includeHiddenZones} = variables || {};

    if (!locales.includes(locale)) {
      return {
        errors: [
          {
            message:
              'Variable $locale of type SupportedLocale! was provided invalid value',
          },
        ],
      };
    }

    const countries = fixtures.countries[locale];

    if (includeHiddenZones) {
      const indiaIndex = countries.data.countries.findIndex(
        (country) => country.code === 'IN',
      );

      // localized names of DH in the supported languages
      const dhNames = {
        // CLDR doesn't translate this name for AF, so use the EN fallback
        en: 'Dadra and Nagar Haveli and Daman and Diu',
        fr: 'Dadra et Nagar Haveli et Daman et Diu',
        ja: 'ダードラー及びナガル・ハヴェーリー及びダマン及びディーウ連邦直轄領',
      };

      countries.data.countries[indiaIndex].zones.push({
        name: dhNames[locale] || dhNames.en,
        code: 'DH',
      });
    }

    if (operationName === 'countries') return countries;

    const country = countries.data.countries.find(
      ({code}) => code === countryCode,
    );

    if (!country)
      return {
        errors: [
          {
            message:
              'Variable $countryCode of type SupportedCountry! was provided invalid value',
          },
        ],
      };
    return {
      data: {
        country,
      },
    };
  };
}
