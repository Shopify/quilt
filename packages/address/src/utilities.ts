import {Address, FieldName, Country, Zone} from './types';

const FIELD_REGEXP = /({\w+})/g;
export const FIELDS_MAPPING: {
  [key: string]: FieldName;
} = {
  '{firstName}': FieldName.FirstName,
  '{lastName}': FieldName.LastName,
  '{country}': FieldName.Country,
  '{city}': FieldName.City,
  '{zip}': FieldName.PostalCode,
  '{province}': FieldName.Zone,
  '{address1}': FieldName.Address1,
  '{address2}': FieldName.Address2,
  '{phone}': FieldName.Phone,
  '{company}': FieldName.Company,
};

/*
 * Returns empty string if all replacement fields are empty.
 */
export function renderLineTemplate(
  country: Country,
  template: string,
  address: Address,
): string {
  const result = template.match(FIELD_REGEXP);
  let line = template;

  if (!result) {
    return '';
  }

  let lineIsEmpty = true;
  result.forEach(key => {
    const addressKey = key.replace('{', '').replace('}', '') as FieldName;

    if (address[addressKey]) {
      lineIsEmpty = false;
    }

    switch (addressKey) {
      case FieldName.Country:
        line = line.replace(`{${FieldName.Country}}`, country.name);
        break;
      case FieldName.Zone:
        line = line.replace(
          `{${FieldName.Zone}}`,
          address.province ? getZone(country.zones, address.province).name : '',
        );
        break;
      default:
        line = line.replace(key, address[addressKey] || '');
        break;
    }
  });
  if (lineIsEmpty) {
    return '';
  } else {
    return line.trim().replace('  ', ' ');
  }
}

function getZone(zones: Zone[], zoneCode: string): Zone {
  return (
    zones.find(zone => zone.code === zoneCode) || {
      name: '',
      code: '',
    }
  );
}
