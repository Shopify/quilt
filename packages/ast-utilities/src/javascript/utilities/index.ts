import {Statement, isStatement} from '@babel/types';
import template from '@babel/template';

export function astFrom(code: string): Statement | undefined {
  const result = template.statement.ast(code, {
    plugins: ['typescript', 'jsx'],
  });

  if (!isStatement(result)) {
    return;
  }

  return result;
}

export function compose(...visitors: any[]) {
  const visitor = visitors.reduce((combinedVisitors, currentVisitor) => {
    return {...combinedVisitors, ...currentVisitor};
  }, {});

  return [
    (function combinedVisitor() {
      return function () {
        return {
          visitor,
        };
      };
    })(),
  ];
}
