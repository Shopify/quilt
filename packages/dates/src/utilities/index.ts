export {formatDate, memoizedGetDateTimeFormat} from './formatDate';
export {getIanaTimeZone} from './timezone';
