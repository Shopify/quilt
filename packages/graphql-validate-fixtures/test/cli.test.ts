/* eslint-disable @shopify/jest/no-snapshots */
import {resolve} from 'path';
import {execSync} from 'child_process';

import {stripFullFilePaths} from '../../../tests/utilities';

const scriptPath = resolve(__dirname, '../bin/graphql-validate-fixtures');
const rootFixturePath = resolve(__dirname, 'fixtures');

// Skipping for now because it fails in CI for no apparent reason :(
// eslint-disable-next-line jest/no-disabled-tests
describe.skip('cli', () => {
  it('succeeds when there are no fixture errors', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('all-clear')),
    ).not.toThrow();
  });

  it('succeeds for a multi-project when there are no fixture errors', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('multi-project')),
    ).not.toThrow();
  });

  it('fails when there are ambiguous fixture names', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('ambiguous-operation')),
    ).toThrowErrorMatchingSnapshot();
  });

  it('fails when a fixture matches no operations', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('missing-operation')),
    ).toThrowErrorMatchingSnapshot();
  });

  it('fails when there are fixture errors', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('fixture-errors')),
    ).toThrowErrorMatchingSnapshot();
  });

  it('fails when there are syntax errors', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('fixture-invalid')),
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      exec(cliCommandForFixtureDirectory('malformed-query')),
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      exec(cliCommandForFixtureDirectory('missing-schema')),
    ).toThrowErrorMatchingSnapshot();
  });

  it('hides passing fixtures by default', () => {
    expect(
      execSync(cliCommandForFixtureDirectory('all-clear')).toString(),
    ).toMatchSnapshot();
  });

  it('shows passes fixtures when the show-passes flag is true', () => {
    const showPasses = true;
    expect(
      execSync(
        cliCommandForFixtureDirectory('all-clear', showPasses),
      ).toString(),
    ).toMatchSnapshot();
  });
});

function exec(command: string) {
  try {
    const result = execSync(command);
    return stripFullFilePaths(result);
  } catch (error) {
    error.message = stripFullFilePaths(error.message);
    throw error;
  }
}

function cliCommandForFixtureDirectory(fixture: string, showPasses = false) {
  const fixtureDirectory = resolve(rootFixturePath, fixture);
  return [
    scriptPath,
    `'${resolve(fixtureDirectory, 'fixtures/**/*.json')}'`,
    `--cwd '${fixtureDirectory}'`,
    showPasses ? `--show-passes` : '',
  ]
    .join(' ')
    .trim();
}

/* eslint-enable @shopify/jest/no-snapshots */
