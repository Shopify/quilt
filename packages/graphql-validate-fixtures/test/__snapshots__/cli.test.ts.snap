// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli fails when a fixture matches no operations 1`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/missing-operation/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/missing-operation'"`;

exports[`cli fails when there are ambiguous fixture names 1`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/ambiguous-operation/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/ambiguous-operation'"`;

exports[`cli fails when there are fixture errors 1`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/fixture-errors/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/fixture-errors'"`;

exports[`cli fails when there are syntax errors 1`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/fixture-invalid/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/fixture-invalid'"`;

exports[`cli fails when there are syntax errors 2`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/malformed-query/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/malformed-query'"`;

exports[`cli fails when there are syntax errors 3`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/missing-schema/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/missing-schema'"`;

exports[`cli hides passing fixtures by default 1`] = `
"

[1m[32m3 passed[39m[22m
"
`;

exports[`cli shows passes fixtures when the show-passes flag is true 1`] = `
"
[7m[1m[32m PASS [39m[22m[27m [2mfixtures/[22m[1manother-another-fixture.json[22m
[7m[1m[32m PASS [39m[22m[27m [2mfixtures/[22m[1manother-fixture.json[22m
[7m[1m[32m PASS [39m[22m[27m [2mfixtures/MyQuery/[22m[1mfixture.json[22m

[1m[32m3 passed[39m[22m
"
`;
